hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+        pos += count
+        bb.put(outputStream.toByteArray)
+        bb.flip()
+
+        assert(bb.position() == 0 && bb.remaining() > 0)
+      }
+
+      override def seek(newPos: Long): Unit = {
+        bb.clear()
+        bb.limit(0)
+        pos = newPos
+      }
+    }
+
+    new WrappedSeekableDataInputStream(is)
+  }
+
+  def createNoCompression(filename: String): PositionedDataOutputStream = {
+    val appendClient: AppendBlobClient = getBlobClient(filename).getAppendBlobClient
+    if (!appendClient.exists()) appendClient.create()
+
+    val os: PositionedOutputStream = new FSPositionedOutputStream {
+      private[this] val client: AppendBlobClient = appendClient
+
+      override def flush(): Unit = {
+        bb.flip()
+
+        if (bb.limit() > 0) {
+          var toWrite: mutable.MutableList[Byte] = mutable.MutableList()
+          var i = 0
+          while (i < bb.remaining()) {
+            toWrite += bb.get(i)
+            i += 1
+          }
+          client.appendBlock(new ByteArrayInputStream(toWrite.toArray), toWrite.length)
+        }
+
+        bb.clear()
+      }
+
+      override def close(): Unit = {
+        if (!closed) {
+          flush()
+          closed = true
+        }
+      }
+    }
+
+    new WrappedPositionedDataOutputStream(os)
+  }
+
+  def delete(filename: String, recursive: Boolean): Unit = {
+    if (this.exists(filename)) {
+      val (account, container, path) = getAccountContainerPath(filename)
+      val blobClient: BlobClient = this.getBlobClient(filename)
+
+      val fileStatus = this.fileStatus(filename)
+
+      if (recursive) {
+        val blobContainerClient = this.getContainerClient(filename)
+
+        val directoryContents = blobContainerClient.listBlobsByHierarchy(dropTrailingSlash(path) + "/")
+        var seenBlobItems: Set[String] = Set()
+        val iter = directoryContents.iterator()
+
+        def deleteRecursively(): Unit = {
+          while (iter.hasNext) {
+            val blobItem: BlobItem = iter.next
+            if (seenBlobItems.contains(blobItem.getName)) return
+            seenBlobItems += blobItem.getName
+            val blobFileName = s"hail-az://$account/$container/${blobItem.getName}"
+            // recursively delete all files/subfolders
+            this.delete(blobFileName, recursive)
+          }
+        }
+        deleteRecursively()
+
+        if (fileStatus.isFile && blobClient.exists()) blobClient.delete()
+      }
+      else {
+          if (fileStatus.isDirectory) {
+            throw new IllegalArgumentException("Cannot non-recursively delete a directory")
+          }
+          if (blobClient.exists()) blobClient.delete()
+      }
+    }
+  }
+
+  def listStatus(filename: String): Array[FileStatus] = {
+    val (account, container, path) = getAccountContainerPath(filename)
+    val blobContainerClient: BlobContainerClient = getContainerClient(filename)
+
+    val statusList: mutable.MutableList[FileStatus] = mutable.MutableList()
```

Comment:
Let's use an `ArrayBuffer` here instead of a `MutableList`, as the former is more like a java `ArrayList` and `MutableList` uses a linked list underneath.